### DISCUSSION TITLE
PyScript reactive framework

### DISCUSSION THREAD

**Discussion by david-ragazzi** [pyscript/pyscript#2142] [Proposals]
Hello everyone, 

I know that PyScript does not yet have reactive features present in React, Vue, Angular, etc, but if you plan do this in the future, here we go a syntax model that could be simple, intuitive and with a smooth learning curve.

This model for PyScript is inspired by syntax of other technologies and that are well accepted by their communities.

Below it's an example of what a code with this model would look like:

```
<script>
    var colors = ['red', 'orange', 'yellow', 'green', 'blue', 'indigo', 'violet']
    var selected = colors[0]

    def selectColor(event):
        color = event.value	  
        selected = color

</script>

<template>
    <h1 style="color: {selected}">Pick a colour</h1>
    <div>
        {% for color in colors %}
            <button
                aria-current={selected === color}
                aria-label={color}
                style="background: {color}"
                onclick={selectColor}
            >{i + 1}</button>
        {% endfor %}
    </div>
</template>

<style>
    button {
        aspect-ratio: 1;
        border-radius: 50%;
        background: var(--color, #fff);
        transform: translate(-2px,-2px);
        filter: drop-shadow(2px 2px 3px rgba(0,0,0,0.2));
        transition: all 0.1s;
    }
</style>
```

![colors](https://github.com/user-attachments/assets/1d78fa5e-b732-451c-a18d-4d73cce21533)

In this model, the code organization is very similar to Svelte and Vue.js, where there is a section for the code (`<script>`), another for the HTML (`<template>`), and another for the styles (`<style>`). These sections would not follow a defined order; the `<style>` section could be moved above all the others, for example.

The states of the variables would be stored implicitly and could be referenced directly by their names, as in Svelte:

`<h1 style="color: {selected}">From the {num_colors} colors bellow pick one:</h1>`

In React we would need:

`colors, setColors = useState(['red', 'orange', 'yellow', 'green', 'blue', 'indigo', 'violet'])`

In Svelte, it's just use `let`, and this model, we would use `var`:

`var colors = ['red', 'orange', 'yellow', 'green', 'blue', 'indigo', 'violet']`

Regarding the template language, Svelte has the [runes](https://svelte.dev/blog/runes), but I think the syntax has too many special characters. Therefore, a really simple syntax that I would suggest would be Django's. Where, for example, a `for` ends with an `endfor`, and an `if` with `endif`:

```
{% if selected == color %}
{% else %}
{% endif %}
```

In Svelte, this would be:

```
{$if selected == color}
{:else}
{/if}
```

That is, 3 special characters (`$`, `:`, `/`), whereas in Django only `%` is used. Django's template language is cleaner, including the end prefix to signal the end of the block.

Well, this is just a suggestion if you want to consider it.

Ps: the `<template>` section could be optional for those who want to keep their code organized. In Svelte, you would use `<main>`, but since it is not mandatory, you can also leave the HTML tags scattered throughout the code.

**Comment by WebReflection** [comment]
There is a native *signals* proposal waiting to land on the JS (hence *DOM*) side of affairs ... runes, Solid Signals, React hooks, they all boil down to the same concept and I think we should wait for the standard to catch up and come out.

I wrote already many reactive libraries and provided *signals* implementation in various fashions so nothing is new to me, I just would like to avoid re-implementing the same wheel for the X time when I know that a standard wheel is coming soon.

I hope this answer makes sense for the time being.

**Reply by david-ragazzi** [reply]
Yes, it does make senses. I agree it's better use the native support (when available). I didn't know this (nice) proposal.

**Comment by kkinder** [comment]
It may (or may not) be of interest to you, and my approach was different from how you laid out your example, but I did write a reactive framework on top of PyScript. I called it [PuePy](https://github.com/kkinder/puepy) (Vue + Py, sort of). You can see a really basic counter example on [puepy.dev](https://puepy.dev) or a more comprehensive example [here](https://expenselemur.com).

It's pure Python, has a (hopefully) pretty minimal codebase, and it works with both Pyodide and Micropython runtimes.

**Reply by david-ragazzi** [reply]
I really liked it. As someone that already code desktop apps with PyQT, it sounds a litte familiar to me, as you write the UI in python code itself, not in a section <script> in a html file.

**Reply by kkinder** [reply]
Yeah that's the idea. I personally find the pure python syntax a bit less verbose and there's no need for figuring out a template system that works in micropython. Plus there's no sea of closing tags to match up.

**Reply by david-ragazzi** [reply]
Man, I'm reading again the github readme. You simply abstracted all the web paraphernalia into a desktop-like application. The learning curve and productivity for a Python developer without much frontend experience becomes ridiculously smooth. I would like this to be publicized more to popularize this concept. Have you been doing this on reddit or other python communities like Django?
